drivers/sensor/ti/ina7xx/ina7xx.c

```diff
@@ -0,0 +1,389 @@
+		if (rc < 0) {
+			LOG_ERR("Error reading diagnostic flags");
+			return rc;
+		}
+
+		if (IS_BIT_SET(tmp, INA7xx_FLAG_CNVRF)) {
+			conversion_complete = true;
+			break;
+		}
+	}
+
+	if (!conversion_complete) {
+		LOG_ERR("Measurement timed out");
+		return -EIO;
+	}
+
+	return rc;
+}
+
+static int ina7xx_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct ina7xx_config *cfg = dev->config;
+	struct ina7xx_data *data = dev->data;
+	uint16_t tmp;
+	int rc = 0;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_VOLTAGE && chan != SENSOR_CHAN_POWER &&
+	    chan != SENSOR_CHAN_DIE_TEMP && chan != SENSOR_CHAN_CURRENT) {
+		return -ENOTSUP;
+	}
+
+	if (cfg->mode != INA7xx_MODE_CONTI) {
+		ina7xx_trigger_measurement(dev);
+	}
+
+	if (chan == SENSOR_CHAN_ALL) {
+		data->valid_readings_mask = 0;
+	}
+
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_VOLTAGE) {
+
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_VOLTAGE_BIT, false);
+		rc = ina7xx_read16(dev, INA7xx_REG_BUS_VOLTAGE, &data->v_bus);
+		if (rc < 0) {
+			LOG_ERR("Error reading bus voltage");
+			return rc;
+		}
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_VOLTAGE_BIT, true);
+	}
+
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_POWER) {
+
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_POWER_BIT, false);
+		rc = ina7xx_read24(dev, INA7xx_REG_POWER, &data->power);
+		if (rc < 0) {
+			LOG_ERR("Error reading power register");
+			return rc;
+		}
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_POWER_BIT, true);
+	}
+
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_CURRENT) {
+
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_CURRENT_BIT, false);
+		rc = ina7xx_read16(dev, INA7xx_REG_CURRENT, &data->current);
+		if (rc < 0) {
+			LOG_ERR("Error reading current register");
+			return rc;
+		}
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_CURRENT_BIT, true);
+	}
+
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_DIE_TEMP) {
+
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_DIE_TEMP_BIT, false);
+		rc = ina7xx_read16(dev, INA7xx_REG_DIE_TEMP, &tmp);
+		if (rc < 0) {
+			LOG_ERR("Error reading temperature register");
+			return rc;
+		}
+		WRITE_BIT(data->valid_readings_mask, INA7xx_READ_DIE_TEMP_BIT, true);
+		data->die_temp = (tmp >> INA7xx_REG_DIE_TEMP_SHIFT);
+	}
+
+	return rc;
+}
+
+static int ina7xx_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct ina7xx_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_VOLTAGE:
+		if (IS_BIT_SET(data->valid_readings_mask, INA7xx_READ_VOLTAGE_BIT)) {
+			sensor_value_from_micro(val,
+						(int64_t)(data->v_bus * INA7xx_BUS_VOLTAGE_MUL_UV));
+		} else {
+			LOG_WRN("Last fetch did not include a voltage reading");
+			return -ENODATA;
+		}
+		break;
+	case SENSOR_CHAN_DIE_TEMP:
+		if (IS_BIT_SET(data->valid_readings_mask, INA7xx_READ_DIE_TEMP_BIT)) {
+			sensor_value_from_micro(val,
```

Comment:
This should be fixed for negative die temperatures.
```suggestion
			int16_t signed_die_temp = (int16_t)data->die_temp;
			signed_die_temp /= 16; /* Remove RESERVED bits 3-0 */
			sensor_value_from_micro(val, signed_die_temp * INA7xx_TEMP_SCALE_MG);
```
Note if the above is used it also requires the shift operation INA7xx_REG_DIE_TEMP_SHIFT to be removed (line 240). It's now achieved with the divide by 16.